File: SignalR.Samples/Hubs/DemoHub/DemoHub.js
$(function () {
    var demo = $.connection.demo;

    demo.invoke = function (index) {
        $('#msg').append('<li>' + index + ' client state index ->' + this.index + '</li>');
    };

    demo.signal = function (id) {
        $('#dynamicTask').html('The dynamic task! ' + id);
    };

    demo.fromArbitraryCode = function (value) {
        $('#arbitraryCode').html('Sending ' + value + ' from arbitrary code without the hub itself!');
    };

    demo.groupAdded = function () {
        $('#groupAdded').append('Group Added');
    };

    demo.errorInCallback = function () {
        var o = null;
        o.doIt();
    };

    $.connection.hub.start(function () {
        demo.doSomethingAndCallError();

        demo.getValue(function (value) {
            $('#value').html('The value is ' + value + ' after 5 seconds');
        });

        var p = {
            Name: "Foo",
            Age: 20,
            Address: { Street: "One Microsoft Way", Zip: "98052" }
        };

        demo.complexType(p, function () {
            $('#value').html('Complex Type ->' + window.JSON.stringify(this.person));
        });

        demo.multipleCalls();

        demo.simpleArray([5, 5, 6]).done(function () {
            $('#simpleArray').html('Simple array works!');
        });

        demo.complexArray([p, p, p]).done(function () {
            $('#complexArray').html('Complex array works!');
        });

        demo.dynamicTask().fail(function () {
            $('#dynamicTask').html('The dynamic task failed :(');
        });

        demo.plainTask().done(function () {
            $('#plainTask').html('Plain Task Result');
        });

        demo.genericTaskTypedAsPlain().done(function (result) {
            $('#genericTaskTypedAsPlain').html(result);
        });

        demo.taskWithException().fail(function (e) {
            $('#taskWithException').html(e);
        });

        demo.addToGroups();
    });
});

File: SignalR.Samples/Hubs/DemoHub/DemoHub.cs
using System;
using System.Threading;
using System.Threading.Tasks;
using SignalR.Hubs;

namespace SignalR.Samples.Hubs.DemoHub
{
    [HubName("demo")]
    public class DemoHub : Hub
    {
        public Task<int> GetValue()
        {
            return Task.Factory.StartNew(() =>
            {
                Thread.Sleep(5000);
                return 10;
            });
        }

        public void AddToGroups()
        {
            AddToGroup("foo");
            AddToGroup("bar");
            Caller.groupAdded();
        }

        public void DoSomethingAndCallError()
        {
            Clients.errorInCallback();
        }

        public Task DynamicTask()
        {
            return Clients.signal(Guid.NewGuid());
        }

        public Task PlainTask()
        {
            return Task.Factory.StartNew(() =>
            {
                Thread.Sleep(500);
            });
        }

        public Task GenericTaskTypedAsPlain()
        {
            return Task.Factory.StartNew(() =>
            {
                return 2 + 2;
            });
        }

        public Task TaskWithException()
        {
            return Task.Factory.StartNew(() =>
            {
                throw new Exception();
            });
        }

        public void SimpleArray(int[] nums)
        {
            foreach (var n in nums)
            {
            }
        }

        public void ComplexArray(Person[] people)
        {

        }

        public Person ComplexType(Person p)
        {
            Caller.person = p;
            return p;
        }

        public void MultipleCalls()
        {
            for (int i = 0; i < 10; i++)
            {
                Caller.index = i + 1;
                Caller.invoke(i);
                Thread.Sleep(1000);
            }
        }

        public class Person
        {
            public string Name { get; set; }
            public int Age { get; set; }
            public Address Address { get; set; }
        }

        public class Address
        {
            public string Street { get; set; }
            public string Zip { get; set; }
        }
    }
}